File: src/infrastructure/provider/openai.ts
/**
 * OpenAI 兼容 Provider — 支持所有 OpenAI 协议兼容的 API
 * (OpenAI, DeepSeek, Moonshot, Zhipu, Ollama, 自定义等)
 */
import type { Message, ProviderConfig, StreamCallback } from '@/types'
import type { ChatRequest, IProvider } from './types'
import {
  apiFetch,
  getResponseReader,
  readStreamLines,
  extractJsonData,
} from './http-utils'
import { v4 } from '@/utils/uuid'
import { getLastSummaryMessageIndex } from '@/types'
import { processVisionContent } from './visionInject'

export class OpenAiProvider implements IProvider {
  readonly name: string
  private apiKey: string
  private baseUrl: string
  private get headers() {
    return {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${this.apiKey}`,
    }
  }

  constructor(name: string, apiKey: string, baseUrl: string) {
    this.name = name
    this.apiKey = apiKey
    this.baseUrl = baseUrl.replace(/\/+$/, '')
  }
  async validateApiKey(config: ProviderConfig): Promise<boolean> {
    if (config.models.length === 0) {
      throw new Error('请先为该 Provider 配置至少一个模型')
    }
    try {
      const res = await fetch(`${config.baseUrl}/chat/completions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${config.apiKey}`,
        },
        body: JSON.stringify({
          model: config.models[0],
          messages: [
            { role: 'user', content: [{ type: 'text', text: 'ping' }] },
          ],
          max_tokens: 1,
        }),
      })
      if (!res.ok) {
        throw new Error(res.statusText)
      }
      return res.ok
    } catch (e) {
      console.error(e)
      return false
    }
  }

  /** 更新 API 配置 */
  updateConfig(apiKey: string, baseUrl: string): void {
    this.apiKey = apiKey
    this.baseUrl = baseUrl.replace(/\/+$/, '')
  }

  async listModels(): Promise<string[]> {
    const res = await apiFetch({
      url: `${this.baseUrl}/models`,
      method: 'GET',
      headers: { Authorization: `Bearer ${this.apiKey}` },
      providerName: this.name,
    })
    const data = await res.json()
    return (data.data || []).map((m: any) => m.id)
  }

  async chat(request: ChatRequest, signal?: AbortSignal): Promise<Message> {
    const body = this.buildRequest(request)
    body.stream = false

    const res = await apiFetch({
      url: `${this.baseUrl}/chat/completions`,
      headers: this.headers,
      body: JSON.stringify(body),
      signal,
      providerName: this.name,
    })

    const data = await res.json()
    return this.parseResponse(data)
  }

  async chatStream(
    request: ChatRequest,
    callback: StreamCallback,
    signal?: AbortSignal,
  ): Promise<void> {
    const body = this.buildRequest(request)
    body.stream = true

    let res: Response
    try {
      res = await apiFetch({
        url: `${this.baseUrl}/chat/completions`,
        headers: this.headers,
        body: JSON.stringify(body),
        signal,
        providerName: this.name,
      })
    } catch (e: any) {
      callback({ type: 'error', error: e.message })
      return
    }

    let reader: ReadableStreamDefaultReader<Uint8Array>
    try {
      reader = getResponseReader(res)
    } catch (e: any) {
      callback({ type: 'error', error: e.message })
      return
    }

    const decoder = new TextDecoder()
    let reasoningContent = ''
    let lastUsage:
      | { promptTokens: number; completionTokens: number; totalTokens: number }
      | undefined
    const toolCallsAccumulator: Map<
      number,
      { id: string; name: string; arguments: string }
    > = new Map()
    let toolCallsFired = false

    try {
      await readStreamLines(
        reader,
        decoder,
        (line) => {
          const parsed = extractJsonData(line)
          if (!parsed) return
          if (parsed.isDone) {
            callback({
              type: 'message_stop',
              reasoningContent,
              usage: lastUsage,
            })
            return
          }

          const chunk = parsed.json
          const delta = chunk.choices?.[0]?.delta

          if (delta?.content) {
            callback({ type: 'text_delta', data: delta.content })
          }
          if (delta?.reasoning_content) {
            reasoningContent += delta.reasoning_content
            callback({
              type: 'reasoning_content_change',
              data: reasoningContent,
            })
          }

          // 处理 tool call delta
          if (delta?.tool_calls) {
            for (const tc of delta.tool_calls) {
              const idx = tc.index ?? 0
              if (!toolCallsAccumulator.has(idx)) {
                toolCallsAccumulator.set(idx, {
                  id: tc.id || '',
                  name: tc.function?.name || '',
                  arguments: tc.function?.arguments || '',
                })
              } else {
                const existing = toolCallsAccumulator.get(idx)!
                if (tc.id) existing.id = tc.id
                if (tc.function?.name) existing.name = tc.function.name
                if (tc.function?.arguments)
                  existing.arguments += tc.function.arguments
              }
            }
          }

          // 记录 usage（流式模式下出现在带有 finish_reason 的最后一个 chunk 中）
          if (chunk.usage) {
            lastUsage = {
              promptTokens: chunk.usage.prompt_tokens ?? 0,
              completionTokens: chunk.usage.completion_tokens ?? 0,
              totalTokens: chunk.usage.total_tokens ?? 0,
            }
          }

          // 检查 finish_reason 是否为 tool_calls（只触发一次）
          if (
            !toolCallsFired &&
            chunk.choices?.[0]?.finish_reason === 'tool_calls'
          ) {
            toolCallsFired = true
            for (const [, tc] of toolCallsAccumulator) {
              callback({
                type: 'tool_use',
                toolUse: {
                  type: 'tool_use',
                  id: tc.id,
                  name: tc.name,
                  input: JSON.parse(tc.arguments || '{}'),
                },
              })
            }
          }
        },
        signal,
      )
    } catch (e: any) {
      if (e?.name !== 'AbortError') {
        callback({ type: 'error', error: e.message })
      }
    }
  }

  buildRequest(request: ChatRequest): any {
    const messages: any[] = []

    // 系统提示词
    if (request.systemPrompt) {
      messages.push({ role: 'system', content: request.systemPrompt })
    }
    const lastSummaryMessageIndex = getLastSummaryMessageIndex(request.messages)
    const requestMessages =
      lastSummaryMessageIndex == -1
        ? request.messages
        : request.messages.slice(lastSummaryMessageIndex)

    // 消息列表
    for (const msg of requestMessages) {
      // summary 角色：转为 user 消息，作为压缩后的历史上下文
      if (msg.role === 'summary') {
        messages.push({
          role: 'user',
          content:
            typeof msg.content === 'string'
              ? `${msg.content}`
              : '' + JSON.stringify(msg.content),
        })
        continue
      }

      const formatted: any = { role: msg.role }

      if (typeof msg.content === 'string') {
        // assistant 带 tool_calls 时，若 content 为空字符串则设为 null（OpenAI strict 模式要求）
        // if (msg.role === 'assistant' && msg.toolCalls?.length && msg.content === '') {
        //   formatted.content = null
        // } else {
        formatted.content = msg.content
        // }
      } else if (Array.isArray(msg.content)) {
        formatted.content = msg.content.map((block) => {
          if (block.type === 'text') return block
          if (block.type === 'image_url') return block
          return block
        })
      }

      // 视觉分析优化：替换 image_url 为分析文本
      const visionBlocks = processVisionContent(msg)
      if (visionBlocks) {
        formatted.content = visionBlocks
      }

      if (msg.toolCalls?.length) {
        formatted.tool_calls = msg.toolCalls.map((tc) => ({
          id: tc.id,
          type: 'function',
          function: { name: tc.name, arguments: JSON.stringify(tc.input) },
        }))
      }

      if (msg.role === 'assistant' && msg.reasoningContent) {
        formatted.reasoning_content = msg.reasoningContent
      }

      if (msg.toolCallId) {
        formatted.tool_call_id = msg.toolCallId
      }

      messages.push(formatted)
    }

    const body: any = {
      model: request.model,
      messages,
      temperature: request.temperature,
      top_p: request.topP,
      max_tokens: request.maxTokens,
      stream: request.stream,
    }

    if (request.tools?.length) {
      body.tools = request.tools.map((t) => ({
        type: 'function',
        function: {
          name: t.name,
          description: t.description,
          parameters: t.parameters,
        },
      }))
    }
    if (request.tool_choice) {
      body.tool_choice = request.tool_choice
    }

    // 推理努力程度（OpenAI o 系列模型：low / medium / high）
    if (request.reasoningEffort) {
      body.reasoning_effort = request.reasoningEffort
    }

    return body
  }

  private parseResponse(data: any): Message {
    const choice = data.choices?.[0]
    const msg = choice?.message || choice?.delta || {}

    const message: Message = {
      id: v4(),
      role: 'assistant',
      content: msg.content || '',
      timestamp: Date.now(),
    }

    if (msg.tool_calls?.length) {
      message.toolCalls = msg.tool_calls.map((tc: any) => ({
        type: 'tool_use' as const,
        id: tc.id,
        name: tc.function.name,
        input: JSON.parse(tc.function.arguments || '{}'),
      }))
    }

    if (data.usage) {
      message.usage = {
        promptTokens: data.usage.prompt_tokens,
        completionTokens: data.usage.completion_tokens,
        totalTokens: data.usage.total_tokens,
      }
    }

    // 保留 reasoning_content（DeepSeek thinking 模式）
    if (msg.reasoning_content) {
      message.reasoningContent = msg.reasoning_content
    }

    return message
  }
}

File: src-tauri/src/agent/provider/openai.rs
//! OpenAI 兼容协议 Provider — OpenAI / DeepSeek / Moonshot / Ollama / 自定义 Base URL
//!
//! 移植自 TS `src/infrastructure/provider/openai.ts`（铁律 1：双引擎同语义）。

use super::blocks::{openai_content, process_vision_content, slice_messages, text_of_content};
use super::sse::read_sse_lines;
use super::Provider;
use super::super::cancellation::CancellationToken;
use super::super::types::{ChatRequest, Message, StreamEvent, TokenUsage, ToolUseContent};
use async_trait::async_trait;
use serde_json::{json, Value};

/// 从 OpenAI 兼容的 usage 里取出**缓存命中**的输入量（与 TS `openai.ts::cachedTokensFromUsage` 对齐）。
///
/// 不取的话账本里的缓存 token 永远是 0：OpenAI 把缓存命中算在 `prompt_tokens` 里，
/// `total - prompt - completion` 恒等于 0，缓存价永远用不上。
///
/// - OpenAI：`prompt_tokens_details.cached_tokens`
/// - DeepSeek：顶层 `prompt_cache_hit_tokens`
/// - 其它兼容实现多数不返回 → `None`（账本退回推导值）
fn openai_cached_tokens(usage: &Value) -> Option<i64> {
    let n = usage
        .get("prompt_tokens_details")
        .and_then(|d| d.get("cached_tokens"))
        .and_then(Value::as_i64)
        .or_else(|| usage.get("prompt_cache_hit_tokens").and_then(Value::as_i64))?;
    if n > 0 {
        Some(n)
    } else {
        None
    }
}

// ==================== OpenAI 兼容 Provider ====================

pub struct NativeOpenAiProvider {
    api_key: String,
    base_url: String,
    http: reqwest::Client,
}

impl NativeOpenAiProvider {
    pub fn new(_name: &str, api_key: &str, base_url: &str) -> Self {
        Self {
            api_key: api_key.to_string(),
            base_url: base_url.trim_end_matches('/').to_string(),
            http: reqwest::Client::new(),
        }
    }

    fn headers(&self) -> reqwest::header::HeaderMap {
        let mut h = reqwest::header::HeaderMap::new();
        h.insert(
            reqwest::header::CONTENT_TYPE,
            reqwest::header::HeaderValue::from_static("application/json"),
        );
        h.insert(
            reqwest::header::AUTHORIZATION,
            reqwest::header::HeaderValue::from_str(&format!("Bearer {}", self.api_key))
                .unwrap_or_else(|_| reqwest::header::HeaderValue::from_static("")),
        );
        h
    }

    /// 构建 OpenAI 兼容请求体（移植 openai.ts buildRequest）
    pub(super) fn build_request(&self, request: &ChatRequest) -> Value {
        let mut messages: Vec<Value> = Vec::new();

        if let Some(sp) = &request.system_prompt {
            messages.push(json!({ "role": "system", "content": sp }));
        }

        let request_messages = slice_messages(&request.messages);
        for msg in request_messages {
            // summary / feedback 角色：转为 user 消息
            if msg.role == "summary" || msg.role == "feedback" {
                messages.push(json!({
                    "role": "user",
                    "content": text_of_content(&msg.content),
                }));
                continue;
            }

            let mut formatted = serde_json::Map::new();
            formatted.insert("role".into(), Value::String(msg.role.clone()));
            // 本地图片伪视觉分析：将图片替换为分析文本（纯文本模型不支持 image 类型）
            // 两条路径都要过块降级（file / quote → 文本），否则自定义块会直接漏给 API
            let content = match process_vision_content(msg) {
                Some(processed) => openai_content(msg, &processed),
                None => openai_content(msg, &msg.content),
            };
            formatted.insert("content".into(), content);

            if let Some(tcs) = &msg.tool_calls {
                if !tcs.is_empty() {
                    let arr: Vec<Value> = tcs
                        .iter()
                        .map(|tc| {
                            json!({
                                "id": tc.id,
                                "type": "function",
                                "function": {
                                    "name": tc.name,
                                    "arguments": serde_json::to_string(&tc.input).unwrap_or_default(),
                                }
                            })
                        })
                        .collect();
                    formatted.insert("tool_calls".into(), Value::Array(arr));
                }
            }

            if msg.role == "assistant" {
                if let Some(rc) = &msg.reasoning_content {
                    formatted.insert("reasoning_content".into(), Value::String(rc.clone()));
                }
            }

            if let Some(tcid) = &msg.tool_call_id {
                formatted.insert("tool_call_id".into(), Value::String(tcid.clone()));
            }

            messages.push(Value::Object(formatted));
        }

        let mut body = json!({
            "model": request.model,
            "messages": messages,
            "temperature": request.temperature,
            "top_p": request.top_p,
            "max_tokens": request.max_tokens,
            "stream": request.stream,
        });

        if !request.tools.is_empty() {
            let tools: Vec<Value> = request
                .tools
                .iter()
                .map(|t| {
                    json!({
                        "type": "function",
                        "function": {
                            "name": t.name,
                            "description": t.description,
                            "parameters": t.parameters,
                        }
                    })
                })
                .collect();
            body["tools"] = Value::Array(tools);
        }
        if !request.tool_choice.is_empty() {
            body["tool_choice"] = Value::String(request.tool_choice.clone());
        }
        if let Some(re) = &request.reasoning_effort {
            body["reasoning_effort"] = Value::String(re.clone());
        }

        body
    }

    /// 解析非流式响应（移植 openai.ts parseResponse）
    fn parse_response(&self, data: &Value) -> Message {
        let choice = data
            .get("choices")
            .and_then(|c| c.as_array())
            .and_then(|arr| arr.first())
            .cloned()
            .unwrap_or_default();
        let msg = choice.get("message").or_else(|| choice.get("delta")).cloned().unwrap_or_default();

        let mut message = Message {
            id: uuid::Uuid::new_v4().to_string(),
            role: "assistant".to_string(),
            content: msg.get("content").and_then(Value::as_str).unwrap_or("").into(),
            tool_calls: None,
            reasoning_content: None,
            tool_call_id: None,
            is_error: None,
            elapsed_ms: None,
            reasoning_elapsed_ms: None,
            ui_data: None,
            timestamp: chrono::Utc::now().timestamp_millis(),
            streaming: None,
            model: None,
            usage: None,
            image_vision_analyze_optimize: None,
            image_vision_analyze_result: None,
        };

        if let Some(tcs) = msg.get("tool_calls").and_then(Value::as_array) {
            if !tcs.is_empty() {
                let parsed: Vec<ToolUseContent> = tcs
                    .iter()
                    .filter_map(|tc| {
                        let id = tc.get("id").and_then(Value::as_str).unwrap_or("").to_string();
                        let name = tc
                            .get("function")
                            .and_then(|f| f.get("name"))
                            .and_then(Value::as_str)
                            .unwrap_or("")
                            .to_string();
                        let args = tc
                            .get("function")
                            .and_then(|f| f.get("arguments"))
                            .and_then(Value::as_str)
                            .unwrap_or("{}");
                        let input = serde_json::from_str(args).unwrap_or(Value::Null);
                        Some(ToolUseContent {
                            type_: "tool_use".into(),
                            id,
                            name,
                            input,
                        })
                    })
                    .collect();
                message.tool_calls = Some(parsed);
            }
        }

        if let Some(usage) = data.get("usage") {
            message.usage = Some(TokenUsage {
                prompt_tokens: usage.get("prompt_tokens").and_then(Value::as_i64).unwrap_or(0),
                completion_tokens: usage.get("completion_tokens").and_then(Value::as_i64).unwrap_or(0),
                total_tokens: usage.get("total_tokens").and_then(Value::as_i64).unwrap_or(0),
                cached_tokens: openai_cached_tokens(usage),
            });
        }

        if let Some(rc) = msg.get("reasoning_content").and_then(Value::as_str) {
            if !rc.is_empty() {
                message.reasoning_content = Some(rc.to_string());
            }
        }

        message
    }
}

#[async_trait]
impl Provider for NativeOpenAiProvider {
    async fn chat(
        &self,
        request: &ChatRequest,
        cancel: &CancellationToken,
    ) -> Result<Message, String> {
        let body = self.build_request(request);
        let url = format!("{}/chat/completions", self.base_url);

        let resp = tokio::select! {
            _ = cancel.cancelled() => return Err("cancelled".into()),
            r = self.http.post(&url).headers(self.headers()).json(&body).send() => r.map_err(|e| format!("API Error: {}", e))?,
        };

        let status = resp.status();
        let text = resp
            .text()
            .await
            .map_err(|e| format!("读取响应失败: {}", e))?;
        if !status.is_success() {
            return Err(format!("API Error ({}): {}", status.as_u16(), text));
        }
        let data: Value =
            serde_json::from_str(&text).map_err(|e| format!("响应解析失败: {}", e))?;
        Ok(self.parse_response(&data))
    }

    async fn chat_stream(
        &self,
        request: &ChatRequest,
        cancel: &CancellationToken,
        on_event: &mut (dyn FnMut(StreamEvent) + Send),
    ) -> Result<(), String> {
        let mut body = self.build_request(request);
        body["stream"] = Value::Bool(true);
        let url = format!("{}/chat/completions", self.base_url);

        let resp = tokio::select! {
            _ = cancel.cancelled() => return Err("cancelled".into()),
            r = self.http.post(&url).headers(self.headers()).json(&body).send() => r.map_err(|e| format!("API Error: {}", e))?,
        };
        let status = resp.status();
        if !status.is_success() {
            let text = resp
                .text()
                .await
                .map_err(|e| format!("读取错误响应失败: {}", e))?;
            return Err(format!("API Error ({}): {}", status.as_u16(), text));
        }

        let mut reasoning_content = String::new();
        let mut last_usage: Option<TokenUsage> = None;
        // index → (id, name, arguments)
        let mut tool_acc: std::collections::HashMap<usize, (String, String, String)> =
            std::collections::HashMap::new();
        let mut tool_fired = false;

        let result = read_sse_lines(resp, cancel, &mut |line: String| {
            let trimmed = line.trim();
            if !trimmed.starts_with("data:") {
                return true;
            }
            let data_str = trimmed[5..].trim().to_string();
            if data_str == "[DONE]" {
                on_event(StreamEvent::MessageStop {
                    reasoning_content: if reasoning_content.is_empty() {
                        None
                    } else {
                        Some(reasoning_content.clone())
                    },
                    usage: last_usage.clone(),
                });
                return true;
            }
            let chunk: Value = match serde_json::from_str(&data_str) {
                Ok(v) => v,
                Err(_) => return true,
            };
            let delta = chunk
                .get("choices")
                .and_then(|c| c.as_array())
                .and_then(|arr| arr.first())
                .and_then(|c| c.get("delta"));

            if let Some(delta) = delta {
                if let Some(content) = delta.get("content").and_then(Value::as_str) {
                    on_event(StreamEvent::TextDelta(content.to_string()));
                }
                if let Some(rc) = delta.get("reasoning_content").and_then(Value::as_str) {
                    reasoning_content.push_str(rc);
                    on_event(StreamEvent::ReasoningContentChange(reasoning_content.clone()));
                }
                if let Some(tcs) = delta.get("tool_calls").and_then(Value::as_array) {
                    for tc in tcs {
                        let idx = tc.get("index").and_then(Value::as_u64).unwrap_or(0) as usize;
                        let entry = tool_acc
                            .entry(idx)
                            .or_insert_with(|| (String::new(), String::new(), String::new()));
                        if let Some(id) = tc.get("id").and_then(Value::as_str).filter(|s| !s.is_empty()) {
                            entry.0 = id.to_string();
                        }
                        if let Some(name) = tc
                            .get("function")
                            .and_then(|f| f.get("name"))
                            .and_then(Value::as_str)
                            .filter(|s| !s.is_empty())
                        {
                            entry.1 = name.to_string();
                        }
                        if let Some(args) = tc
                            .get("function")
                            .and_then(|f| f.get("arguments"))
                            .and_then(Value::as_str)
                        {
                            entry.2.push_str(args);
                        }
                    }
                }
            }

            if let Some(usage) = chunk.get("usage") {
                last_usage = Some(TokenUsage {
                    prompt_tokens: usage.get("prompt_tokens").and_then(Value::as_i64).unwrap_or(0),
                    completion_tokens: usage.get("completion_tokens").and_then(Value::as_i64).unwrap_or(0),
                    total_tokens: usage.get("total_tokens").and_then(Value::as_i64).unwrap_or(0),
                    cached_tokens: openai_cached_tokens(usage),
                });
            }

            // finish_reason === tool_calls
            if !tool_fired {
                let finish = chunk
                    .get("choices")
                    .and_then(|c| c.as_array())
                    .and_then(|arr| arr.first())
                    .and_then(|c| c.get("finish_reason"))
                    .and_then(Value::as_str);
                if finish == Some("tool_calls") {
                    tool_fired = true;
                    let mut items: Vec<(usize, (String, String, String))> =
                        tool_acc.drain().collect();
                    items.sort_by_key(|(idx, _)| *idx);
                    for (_, (id, name, args)) in items {
                        let input = serde_json::from_str(&args).unwrap_or(Value::Null);
                        on_event(StreamEvent::ToolUse(ToolUseContent {
                            type_: "tool_use".into(),
                            id,
                            name,
                            input,
                        }));
                    }
                }
            }
            true
        })
        .await;

        result
    }
}
